cmake/download_archive.cmake

```diff
@@ -0,0 +1,39 @@
+# Copyright 2021 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set(_download_archive_TEMPORARY_DIR ${CMAKE_BINARY_DIR}/http_archives)
+file(MAKE_DIRECTORY ${_download_archive_TEMPORARY_DIR})
+
+# This is basically Bazel's http_archive.
+# Note that strip_prefix strips the directory path prefix of the extracted
+# archive content, and it may strip multiple directories.
+function(download_archive destination url hash strip_prefix)
```

Comment:
qq: have you investigated whether cmake already has some useful features we could reuse that are sophisticated than `file(DOWNLOAD)`? I want to avoid us reinventing the wheel.

Reply:
There is a `ExternalProject` suite of commands: https://cmake.org/cmake/help/latest/module/ExternalProject.html

To use it, we might need to bump up our minimum CMake version to at least 3.7 (we are on 3.5.1 now). WDYT?

Reply:
I think ExternalProject has been available much earlier than cmake 3.7 (unless you need some specific features that were introduced in 3.7. do you?), but there might be other issues with using ExternalProject.

Generally, the problem with ExternalProject is that it only download the dependencies in the "build" phase, which might be too late in some cases (e.g. when you need the external project to do something in the "configure" phase). This might be ok for the external proto libraries, since they don't have a CMakeLists.txt themselves, but it would not work in some more complex cases (e.g. download UPB dependency and than include it in a build).
IIRC the `ExternalProject` isn't very popular due to its significant limitations.

In cmake 3.11 version, the `FetchContent` was introduced, which is much more flexible and usable as AFAICT, it's now the "recommended" one to use over the `ExternalProject`. 
https://cmake.org/cmake/help/latest/module/FetchContent.html


Quoting from the docs:
> This module enables populating content at configure time via any method supported by the [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html#module:ExternalProject) module. Whereas [ExternalProject_Add()](https://cmake.org/cmake/help/latest/module/ExternalProject.html#command:externalproject_add) downloads at build time, the FetchContent module makes content available immediately, allowing the configure step to use the content in commands like [add_subdirectory()](https://cmake.org/cmake/help/latest/command/add_subdirectory.html#command:add_subdirectory), [include()](https://cmake.org/cmake/help/latest/command/include.html#command:include) or [file()](https://cmake.org/cmake/help/latest/command/file.html#command:file) operations.

See e.g. https://bewagner.net/programming/2020/05/02/cmake-fetchcontent/ for more info.


TL;DR from this is that for now we can keep the current `file(DOWNLOAD ` and maybe in the future we can look into converting our build to use FetchContent (some time in the future, when requiring e.g. cmake 3.14 isn't an issue for us), that has some potential for simplifying our build.

Reply:
As we discussed offline, there are rooms for improvement. We can migrate to the newer `FetchContent` command, but we will need to investigate it's behavior (whether it's in BUILD stage or CONFIGURE stage).